index.d.ts

```diff
@@ -1,55 +1,86 @@
-export default update
-
-declare function update<T>(
-  data: ReadonlyArray<T>,
-  query: ArrayOperators<T>,
-): ReadonlyArray<T>
-
-declare function update<T>(
-  data: ReadonlySet<T>,
-  query: SetOperators<T>,
-): ReadonlySet<T>
-
-declare function update<K, V>(
-  data: ReadonlyMap<K, V>,
-  query: MapOperators<K, V>,
-): ReadonlyMap<K, V>
-
-declare function update<T>(data: T, query: Query<T>): T
-
-type Tree<T> = {[K in keyof T]?: Query<T[K]>}
-export type Query<T> =
-  | Tree<T>
-  | ObjectOperators<T>
-  | ArrayOperators<any>
-  | SetOperators<any>
-
-type ObjectOperators<T> =
-  | {$set: any}
-  | {$toggle: Array<keyof T | number>}
-  | {$unset: Array<keyof T | number>}
-  | {$merge: Partial<T>}
-  | {$apply: (old: T) => T}
-  | ((old: T) => any)
-type ArrayOperators<T> =
-  | {$push: T}
-  | {$unshift: T}
-  | {$splice: Array<[number, number]>}
-  | {[customCommand: string]: any}
-
-type MapOperators<K, V> = {$add: Array<[K, V]>} | {$remove: K[]}
-type SetOperators<T> = {$add: T[]} | {$remove: T[]}
+// TypeScript Version: 2.9
+
+export default update;
+
+// C is the custom commands object. It's defined as a type parameter because users should
+// explicitly specify what their legal extra commands are. This function cannot effectively
+// be typed to automatically accept the default commands as well as arbitrary custom commands.
+// The reason for this is that any such automatic inference will necessarily require the
+// types to be extremely permissive, because we don't know what the custom commands are at
+// call time. So permissive, in fact, that any typo or improper value type in the spec
+// argument will be accepted because they would have to be assumed to be a custom command.
+//
+// Since such permissive-as-to-be-entirely-useless behavior is not desirable, the user must
+// explicitly specify what custom commands are allowed. However, if C is defined as
+// `C extends object` or something similarly simple, Typescript will freely infer any typos
+// or improper value types in the spec (again) as custom commands. Instead, C is wrapped in
+// a "branded" type to prevent any kind of inference, and then that brand is unwrapped and
+// discarded so that the spec the user provides doesn't actually have to include the branded
+// field.
+//
+// Usage with custom commands is as follows:
+//
+//   interface MyCommands {
+//     $foo: string;
+//   }
+//
+//    update<Foo, CustomCommands<MyCommands>>(..., { $foo: "bar" });
+//
+// It is suggested that if you use custom commands frequently, you wrap and re-export a
+// properly-typed version of `update`:
+//
+//   function myUpdate<T>(object: T, spec: Spec<T, CustomCommands<MyCommands>>) {
+//     return update(object, spec);
+//   }
+declare function update<T, C extends CustomCommands<object> = never>(
+  target: T,
+  spec: Spec<T, C>,
+): T;
+
+// See the comment on `update` for an explanation of this type.
+export type CustomCommands<T> = T & { __noInferenceCustomCommandsBrand: any };
```

Comment:
This is somewhat unfortunate but not a dealbreaker, I don't think. I'm not sure what the best place to include the lengthy explanatory comment is (Github comment? Code comment?).